elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchTable.java

```diff
@@ -87,37 +126,216 @@ private static ElasticsearchVersion detectVersion(RestClient client, ObjectMappe
+    ObjectNode parent = query.with(AGGREGATIONS);
+    for (String name: orderedGroupBy) {
+      final String aggName = "g_" + name;
+      fieldMap.put(aggName, name);
+
+      final ObjectNode section = parent.with(aggName);
+      final ObjectNode terms = section.with("terms");
+      terms.put("field", name);
+      terms.set("missing", ElasticsearchJson.MISSING_VALUE); // expose missing terms
+
+      if (fetch != null) {
+        terms.put("size", fetch);
+      }
+
+      sort.stream().filter(e -> e.getKey().equals(name)).findAny().ifPresent(s -> {
+        terms.with("order").put("_key", s.getValue().isDescending() ? "desc" : "asc");
+      });
+
+      parent = section.with(AGGREGATIONS);
+    }
+
+    // simple version for queries like "select count(*), max(col1) from table" (without GROUP BY)
+    if (!(aggregations.stream().allMatch(isCountStar) && groupBy.isEmpty())) {
+      for (Map.Entry<String, String> aggregation : aggregations) {
+        JsonNode value = mapper.readTree("{" + aggregation.getValue()  + "}");
+        parent.set(aggregation.getKey(), value);
+      }
+    }
+
+    // cleanup query. remove empty AGGREGATIONS element (if empty)
+    JsonNode agg = query;
+    while (agg.has(AGGREGATIONS) && agg.get(AGGREGATIONS).elements().hasNext()) {
+      agg = agg.get(AGGREGATIONS);
+    }
+    ((ObjectNode) agg).remove(AGGREGATIONS);
+
+    ObjectNode node = httpRequest(query);
+    ElasticsearchJson.Result res = mapper.treeToValue(node,
+                ElasticsearchJson.Result.class);
+
+
+    final List<Map<String, Object>> result = new ArrayList<>();
+    if (res.aggregations() != null) {
+      // collect values
+      ElasticsearchJson.visitValueNodes(res.aggregations(), m -> {
+        Map<String, Object> newMap = new LinkedHashMap<>();
+        for (String key: m.keySet()) {
+          newMap.put(fieldMap.getOrDefault(key, key), m.get(key));
+        }
+        result.add(newMap);
+      });
+    } else {
+      // probably no group by. add single result
+      result.add(new LinkedHashMap<>());
+    }
+
+    // elastic exposes total number of documents matching a query in "/hits/total" path
+    // this can be used for simple "select count(*) from table"
+    final long total = res.searchHits().total();
+
+    if (groupBy.isEmpty()) {
+      // put totals automatically for count(*) expression(s), unless they contain group by
+      for (String expr : countAll) {
+        result.forEach(m -> m.put(expr, total));
+      }
+    }
+
+    final Function1<ElasticsearchJson.SearchHit, Object> getter =
+        ElasticsearchEnumerators.getter(fields);
+
+    ElasticsearchJson.SearchHits hits =
+        new ElasticsearchJson.SearchHits(total, result.stream()
+            .map(r -> new ElasticsearchJson.SearchHit("_id", r, null))
+            .collect(Collectors.toList()));
+
+    return Linq4j.asEnumerable(hits.hits()).select(getter);
+  }
+
+  private ObjectNode httpRequest(ObjectNode query) throws IOException {
```

Comment:
IMO this should always return `ElasticseachrJson.Result`.

Reply:
fixed